gcp/src/main/java/org/apache/iceberg/gcp/biglake/BigLakeCatalog.java

```diff
@@ -0,0 +1,396 @@
+          String.format("Invalid namespace (too long): %s", namespace));
+    }
+  }
+
+  @Override
+  public List<Namespace> listNamespaces(Namespace namespace) {
+    if (namespace.isEmpty()) {
+      return Streams.stream(client.listDatabases(catalogName))
+          .map(BigLakeCatalog::namespace)
+          .collect(ImmutableList.toImmutableList());
+    }
+
+    // Database namespace does not have nested namespaces.
+    if (namespace.levels().length == 1) {
+      return ImmutableList.of();
+    }
+
+    throw new NoSuchNamespaceException("Invalid namespace: %s", namespace);
+  }
+
+  @Override
+  public boolean dropNamespace(Namespace namespace) {
+    try {
+      if (namespace.isEmpty()) {
+        // Used by `DROP NAMESPACE <catalog>`. Deletes the BLMS catalog linked by Iceberg catalog.
+        client.deleteCatalog(catalogName);
+        LOG.info("Deleted BigLake catalog: {}", catalogName.toString());
+      } else if (namespace.levels().length == 1) {
+        String dbId = databaseId(namespace);
+        validateDatabaseId(dbId, namespace);
+        client.deleteDatabase(databaseName(dbId));
+        // Don't delete the data file folder for safety. It aligns with HMS's default behavior.
+        // To support database or catalog level config controlling file deletion in future.
+      } else {
+        LOG.warn("Invalid namespace (too long): {}", namespace);
+        return false;
+      }
+    } catch (NoSuchNamespaceException e) {
+      LOG.warn("Failed to drop namespace", e);
+      return false;
+    }
+
+    return true;
+  }
+
+  @Override
+  public boolean setProperties(Namespace namespace, Map<String, String> props) {
+    String dbId = databaseId(namespace);
+    validateDatabaseId(dbId, namespace);
+
+    HiveDatabaseOptions.Builder optionsBuilder =
+        loadDatabase(dbId).toBuilder().getHiveOptionsBuilder();
+    props.forEach(optionsBuilder::putParameters);
+    client.updateDatabaseParameters(databaseName(dbId), optionsBuilder.getParametersMap());
+    return true;
+  }
+
+  @Override
+  public boolean removeProperties(Namespace namespace, Set<String> props) {
+    String dbId = databaseId(namespace);
+    validateDatabaseId(dbId, namespace);
+
+    HiveDatabaseOptions.Builder optionsBuilder =
+        loadDatabase(dbId).toBuilder().getHiveOptionsBuilder();
+    props.forEach(optionsBuilder::removeParameters);
+    client.updateDatabaseParameters(databaseName(dbId), optionsBuilder.getParametersMap());
+    return true;
+  }
+
+  @Override
+  public Map<String, String> loadNamespaceMetadata(Namespace namespace) {
+    if (namespace.isEmpty()) {
+      // Calls catalog to check existence. BLMS catalog has no metadata today.
+      client.catalog(catalogName);
+      return ImmutableMap.of();
+    } else if (namespace.levels().length == 1) {
+      String dbId = databaseId(namespace);
+      validateDatabaseId(dbId, namespace);
+      return loadDatabase(dbId).getHiveOptions().getParametersMap();
```

Comment:
Is this parameter map immutable or unmodifiable?

Reply:
It is unmodifiable: https://protobuf.dev/reference/java/java-generated/#map-fields